clang-tools-extra/clang-tidy/modernize/UseIntegerSignComparisonCheck.cpp

```diff
@@ -0,0 +1,192 @@
+  const Expr *RHS = BinaryOp->getRHS()->IgnoreParenImpCasts();
+  if (LHS == nullptr || RHS == nullptr)
+    return;
+
+  const Expr *ExplicitLHS = nullptr;
+  const Expr *ExplicitRHS = nullptr;
+  StringRef ExplicitLhsString, ExplicitLhsRhsString;
+
+  DiagnosticBuilder Diag =
+      diag(BinaryOp->getBeginLoc(),
+           "comparison between 'signed' and 'unsigned' integers");
+  if (BinaryOp->getLHS() == SignedCastExpression) {
+    ExplicitLHS = SignedCastExpression->isPartOfExplicitCast()
+                      ? SignedCastExpression
+                      : nullptr;
+    ExplicitRHS = UnSignedCastExpression->isPartOfExplicitCast()
+                      ? UnSignedCastExpression
+                      : nullptr;
+  } else {
+    ExplicitRHS = SignedCastExpression->isPartOfExplicitCast()
+                      ? SignedCastExpression
+                      : nullptr;
+    ExplicitLHS = UnSignedCastExpression->isPartOfExplicitCast()
+                      ? UnSignedCastExpression
+                      : nullptr;
+  }
+
+  if (!(getLangOpts().CPlusPlus17 && IsQtApplication) &&
+      !getLangOpts().CPlusPlus20)
+    return;
+
+  std::string CmpNamespace;
+  std::string CmpHeader;
+  if (getLangOpts().CPlusPlus20) {
+    CmpNamespace = std::string("std::") + std::string(parseOpCode(OpCode));
+    CmpHeader = std::string("<utility>");
+  } else if (getLangOpts().CPlusPlus17 && IsQtApplication) {
+    CmpNamespace = std::string("q20::") + std::string(parseOpCode(OpCode));
+    CmpHeader = std::string("<QtCore/q20utility.h>");
+  }
+
+  // Use qt-use-integer-sign-comparison when C++17 is available and only for Qt
+  // apps. Prefer modernize-use-integer-sign-comparison when C++20 is available!
+  if (ExplicitLHS) {
+    ExplicitLhsString =
+        Lexer::getSourceText(CharSourceRange::getTokenRange(
+                                 ExplicitLHS->IgnoreCasts()->getSourceRange()),
+                             *Result.SourceManager, getLangOpts());
+    Diag << FixItHint::CreateReplacement(
+        LHS->getSourceRange(),
+        llvm::Twine(CmpNamespace + "(" + ExplicitLhsString).str());
+  } else {
+    Diag << FixItHint::CreateInsertion(LHS->getBeginLoc(),
+                                       llvm::Twine(CmpNamespace + "(").str());
+  }
+  Diag << FixItHint::CreateReplacement(BinaryOp->getOperatorLoc(), ", ");
+  if (ExplicitRHS) {
+    ExplicitLhsRhsString =
+        Lexer::getSourceText(CharSourceRange::getTokenRange(
+                                 ExplicitRHS->IgnoreCasts()->getSourceRange()),
+                             *Result.SourceManager, getLangOpts());
+    Diag << FixItHint::CreateReplacement(
+        RHS->getSourceRange(), llvm::Twine(ExplicitLhsRhsString + ")").str());
+  } else {
+    Diag << FixItHint::CreateInsertion(
+        Lexer::getLocForEndOfToken(
+            Result.SourceManager->getSpellingLoc(RHS->getEndLoc()), 0,
+            *Result.SourceManager, Result.Context->getLangOpts()),
+        ")");
+  }
```

Comment:
This can be simplified some more:

Say we have the case of both sides with explicit casts:

```
static_cast<int>(42) < static_cast<char>(2U)
```

Then we only need to do three replacements and know three ranges.

We already have:

```
static_cast<int>(42) < static_cast<char>(2U)
|------------------| | |--------------------|
          A          B          C
```

and want to end up with

```
std::cmp_less(42, 2U)
```

```
static_cast<int>(42) < static_cast<char>(2U)
|---------------|  |--------------------|  |
        R1                    R2           R3
   std::cmp_less(             ,            )    // the replacements for each range
```

E.g., (partial, untested)
```c++
const auto* LHS = BinOp->>getLHS()->IgnoreParenImpCasts();
auto R1 = SourceRange(LHS->getBeginLoc());
auto R2 = SourceRange(LHS->getEndLoc(), RHS->getBeginLoc());
if (const auto * LHSCast = llvm::dyn_cast<ExplicitCastExpr>(LHS)) {
  const auto * SubExpr = LHSCast->getSubExpr();
  R1 = SourceRange(LHS->getBeginLoc(), SubExpr->getBeginLoc());
  R2.setBegin(SubExpr->getEndLoc().getWithOffset(1));
}
```

And in the case of an operand without an explicit cast, the range `R1` will be like an insertion point.

Reply:
By my tests, llvm::dyn_cast<ExplicitCastExpr> always returned null, so I didn't use it. But I refactored my version using this improvement.